### PR TITLE
[SPARK-38896][CORE][SQL] Use `tryWithResource` release `LevelDB/RocksDBIterator` resources earlier

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
Similar to SPARK-38847, this pr aims to release the `LevelDB/RocksDBIterator` resources earlier by using `tryWithResource`. The main change of this pr as follows:

1. Use Java `tryWithResource` and Spark `Utils.tryWithResource` to recycling `KVStoreIterator` opened by `RocksDB.view(Class<T> type).iterator` and `RocksDB.view(Class<T> type).iterator`
2. Introduce 4 new function for KVUtils(`count|foreach|mapToSeq|size`),  these function will close `KVStoreIterator` in time.

### Why are the changes needed?
Release the `LevelDB/RocksDBIterator` resources earlier

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Pass GA

### CODE REVIEW

**Comment by LuciferYang** [comment]
Test link : https://github.com/LuciferYang/spark/runs/6051872250

File: common/kvstore/src/test/java/org/apache/spark/util/kvstore/RocksDBSuite.java [hunk 1]
```diff
@@ -255,12 +257,15 @@ public void testNegativeIndexValues() throws Exception {
...
+        .stream(Spliterators.spliteratorUnknownSize(iterator, 0), false)
+        .map(e -> e.num)
+        .collect(Collectors.toList());
 
-    assertEquals(expected, results);
+      assertEquals(expected, results);
+    }
   }
 
   @Test
```
Comment (LuciferYang) [inline common/kvstore/src/test/java/org/apache/spark/util/kvstore/RocksDBSuite.java:271]: This pr clears all similar cases except `testCloseRocksDBIterator` in `LevelDB/RocksDBSuite` because this case seems use to check `finalize` mechanism.

After this pr, we can consider cleaning up `finalize()` related code from `LevelDB/RocksDB` because the `finalize` mechanism is not recommended, and it has been identified as `forRemoval=true` after [Java 18](https://openjdk.java.net/jeps/421)

cc @srowen @HyukjinKwon @dongjoon-hyun

File: common/kvstore/src/test/java/org/apache/spark/util/kvstore/LevelDBBenchmark.java [hunk 2]
```diff
@@ -197,9 +197,15 @@ private void iterate(KVStoreView<?> view, String name) throws Exception {
...
-    while (it.hasNext()) {
-      try(Timer.Context ctx = iter.time()) {
-        it.next();
+    try {
+      while (it.hasNext()) {
+        try (Timer.Context ctx = iter.time()) {
+          it.next();
+        }
+      }
+    } finally {
```
Comment (LuciferYang) [inline common/kvstore/src/test/java/org/apache/spark/util/kvstore/LevelDBBenchmark.java:206]: The last `it` is not closed

File: common/kvstore/src/test/java/org/apache/spark/util/kvstore/RocksDBBenchmark.java [hunk 3]
```diff
@@ -196,10 +196,15 @@ private void iterate(KVStoreView<?> view, String name) throws Exception {
...
-    while (it.hasNext()) {
-      try(Timer.Context ctx = iter.time()) {
-        it.next();
+    try {
+      while (it.hasNext()) {
+        try (Timer.Context ctx = iter.time()) {
+          it.next();
+        }
+      }
+    } finally {
```
Comment (LuciferYang) [inline common/kvstore/src/test/java/org/apache/spark/util/kvstore/RocksDBBenchmark.java:205]: The last it is not closed

**Review by srowen — COMMENTED** [review]
I'm not against this, but it adds complexity. Do you have any rough sense of how much it improves things?

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala [hunk 4]
```diff
@@ -325,12 +325,18 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
 
   override def getListing(): Iterator[ApplicationInfo] = {
     // Return the listing in end time descending order.
-    listing.view(classOf[ApplicationInfoWrapper])
+    // SPARK-38896: tryWithResource cannot be used here
+    // because this method needs to return an `Iterator`.
+    val closeableIter = listing.view(classOf[ApplicationInfoWrapper])
```
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:330]: `tryWithResource` cannot be used here because this method needs to return an `Iterator`,  mixed in `java.io.Closeable` so that the caller can close it

**Comment by LuciferYang** [comment]
> I'm not against this, but it adds complexity. Do you have any rough sense of how much it improves things?

Sorry, I want to clarify the main goal of this pr: it is not for the benefit of performance, but to close `KVStoreIterator ` resources opened by `KVStore.view.iterator` in time.

I hope `finalize()`  method can be cleaned up later after this pr because using `finalize` to recycle resources is not a good way( for example, Effective Java Item 7 Avoid finalizers) and `finalize()` is identified as `deprecated` after Java 11 and `forRemoval=true` after Java 18

**Comment by LuciferYang** [comment]
Updated PR description

**Comment by LuciferYang** [comment]
@srowen Java 9 provides [java.lang.ref.Cleaner](https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/ref/Cleaner.java) which can achieve the same purpose as `finalize`, but Spark still need to support Java 8 now. 

Are there any other recommended alternatives?

**Comment by srowen** [comment]
Right, I agree it's good practice, just wondering what we gain from it. For the in-memory case it does nothing. It looks like this releases some RocksDB state for example and that's good to do earlier.

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala [hunk 5]
```diff
@@ -325,12 +325,18 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
...
-      .iterator()
-      .asScala
-      .map(_.toApplicationInfo())
+      .closeableIterator()
+    val dataIter = closeableIter.asScala.map(_.toApplicationInfo())
+    new Iterator[ApplicationInfo] with Closeable {
+      override def hasNext: Boolean = dataIter.hasNext
+      override def next(): ApplicationInfo = dataIter.next()
+      override def close(): Unit = closeableIter.close()
+    }
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:339]: Since not all usages of `getListing()` are draining the iterator, add a `finalize` as well ?

Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:339]: `finalize` or `finally`?

**Comment by LuciferYang** [comment]
> Right, I agree it's good practice, just wondering what we gain from it. For the in-memory case it does nothing. It looks like this releases some RocksDB state for example and that's good to do earlier.

Yes, this pr is to release `LevelDB/RocksDBIterator` earlier and not relate to `InMemoryIterator`

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala (hunk 5 above)
Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:339]: If `finalize`, there is no need to change here. But I don't think it should be used `finalize()` anymore. It's not a good practice, and it has been identified as forRemoval=true after [JEP 421](https://openjdk.java.net/jeps/421)

If use `finally` here to close `LevelDB/RocksDBIterator`, the caller will not be able to consume data.

Comment (srowen) [reply core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:339]: finalize() isn't wrong per se; relying on it can be. Given the context here, I think you would want to include it.

Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:339]: Thank you for your help  @srowen

**Comment by LuciferYang** [comment]
All Test passed: https://github.com/LuciferYang/spark/runs/6071316292

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala (hunk 5 above)
Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:339]: Or we can return
```
KVUtils.mapToSeq(listing.view(classOf[ApplicationInfoWrapper])
      .index("endTime")
      .reverse())(_.toApplicationInfo()).iterator
```
This seems simpler, and the caller doesn't need to care about closing `LevelDB/RocksDBIterator`, but it may use more memory

File: core/src/main/scala/org/apache/spark/status/KVUtils.scala [hunk 6]
```diff
@@ -100,6 +100,33 @@ private[spark] object KVUtils extends Logging {
     }
   }
 
+  /** Counts the number of elements in the KVStoreView which satisfy a predicate. */
+  def count[T](view: KVStoreView[T])(countFunc: T => Boolean): Int = {
```
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/status/KVUtils.scala:104]: @srowen Do you mind if I split the newly introduced functions and related refactoring work into another pr? It seems that the change of this part does not increase the complexity of the code.

Comment (srowen) [reply core/src/main/scala/org/apache/spark/status/KVUtils.scala:104]: I think that's OK

Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/status/KVUtils.scala:104]: OK ~

Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/status/KVUtils.scala:104]: @srowen I separate the code related to these 4 new functions into a new pr: https://github.com/apache/spark/pull/36237/files

**Comment by srowen** [comment]
That seems like an overlapping change - which one of the two is ready for review/merge? I would have imagined it's this PR?

**Comment by LuciferYang** [comment]
> That seems like an overlapping change - which one of the two is ready for review/merge? I would have imagined it's this PR?

OK， let's continue this pr

**Comment by LuciferYang** [comment]
I am ok to continue on this pr

**Comment by srowen** [comment]
OK, which change do you want to go in first? this PR has both changes right now, it appears

**Comment by LuciferYang** [comment]
I close another PR, and I hope to fix similar cases all in one

**Comment by srowen** [comment]
Sorry I am still not clear. Do you want to review and merge this PR, now?

File: core/src/main/scala/org/apache/spark/status/AppStatusStore.scala [hunk 7]
```diff
@@ -757,7 +767,7 @@ private[spark] class AppStatusStore(
         executorLogs,
         AppStatusUtils.schedulerDelay(taskDataOld),
         AppStatusUtils.gettingResultTime(taskDataOld))
-    }.toSeq
+    }.toList
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/status/AppStatusStore.scala:770]: Out of curiosity, why this change?

File: core/src/test/scala/org/apache/spark/deploy/history/HybridStoreSuite.scala [hunk 8]
```diff
@@ -103,6 +104,13 @@ abstract class HybridStoreSuite extends SparkFunSuite with BeforeAndAfter with T
   }
 
   test("test basic iteration") {
+
+    def firstInView[T](view: KVStoreView[T]): T = {
```
Comment (srowen) [inline core/src/test/scala/org/apache/spark/deploy/history/HybridStoreSuite.scala:108]: Do you want to call this "head"?

**Review by srowen — COMMENTED** [review]
Was there anything you wanted to change w.r.t. comment https://github.com/apache/spark/pull/36237/files#r852703387 ? Looking pretty fine

**Comment by LuciferYang** [comment]
> Was there anything you wanted to change w.r.t. comment https://github.com/apache/spark/pull/36237/files#r852703387 ? Looking pretty fine

[f91bd0d](https://github.com/apache/spark/pull/36237/commits/f91bd0d3cf4262095f490996a698393e4de05d2b) fix this

File: core/src/test/scala/org/apache/spark/deploy/history/HybridStoreSuite.scala (hunk 8 above)
Comment (LuciferYang) [reply core/src/test/scala/org/apache/spark/deploy/history/HybridStoreSuite.scala]: [70528b6](https://github.com/apache/spark/pull/36237/commits/70528b6d0af97636845196e4a764fdce6d41e2db) fix this

File: core/src/main/scala/org/apache/spark/status/AppStatusStore.scala (hunk 7 above)
Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/status/AppStatusStore.scala:770]: Let me double check this

Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/status/AppStatusStore.scala]: This change is to make `HistoryServerSuite` passed,  `toSeq` causes `LevelDB/RocksDBIterator` close in advance, while 'toList' does not. 

[de2862e](https://github.com/apache/spark/pull/36237/commits/de2862ec2fd1caec1c74b092510982b287795b58) try to fixed the issue in a different way

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master
